### ISSUE TITLE
large images are rotated when scaled

### ISSUE THREAD

**Issue by troy2914** [icpctools/icpctools#741]
At the ACPC we had images that were 4000x6000,  and running the Resolver against the CDP resulted in the pictures being rotated 90 degrees.   Interestingly, rotating the images 90 degrees before giving to the Resolver resulted in no extra rotation (still 90 degrees).

As a work around we reduced the size of the image by 20% in each direction.  That resulted in the Resolver displaying the image properly,  and of note that also mean the image did not need to be resized.

This was using pre-release Resolver v2.5.868

**Comment by nickygerritsen** [comment]
Could we get such an image? I think it’s specific to some files (and rotating or reducing the size changes the file)

**Comment by troy2914** [comment]
JohnC shared a zip of the image with Nicky and Tim on slack.

**Comment by nickygerritsen** [comment]
Which indeed confirmed my hunch: the EXIF info contains rotation info, which the pres clients should use.

**Comment by deboer-tim** [comment]
Can you confirm the size of the display (or attach the logs)? There is some special code to handle downsizing special cases (the built-in algorithms are either too slow or produce artifacts) and I'm guessing it's only failing to keep the exif in one of these paths.

**Comment by clevengr** [comment]
The display was my ThinkPad X1:  LENOVO_MT_20QV_BU_Think_FM_ThinkPad X1 Extreme 2nd, with the display set to 1920x1080.  The attached log confirms that.  We drove the presentation screen directly from my laptop's HDMI output, with no CDS or Presenter Client involved.  (Note that I leave my laptop on home time, so the log times are California -- 10 hours earlier than the actual local time the event took place.)
[resolver_22.12.04_09.21.46.log](https://github.com/icpctools/icpctools/files/10169788/resolver_22.12.04_09.21.46.log)

**Comment by johnbrvc** [comment]
> 

Yes, we saw the special code and that's what gave us the idea of "pre-shrinking" the images using "mspaint", which, as you now know, worked.

**Comment by deboer-tim** [comment]
I confirmed the issue is lost exif orientation metadata, and that it's lost on ImageIO import. i.e. scaling was a red herring, the orientation is already lost on import and the image would have been displayed wrong. If I try to save the original image you get the rotated image w/o exif orientation metadata. I'm pretty sure mspaint was also rotating the image and resetting the orientation whether or not you resized it.

I've found several third-party libraries that can read exif data but all require an extra library, code, and reading the image 'twice' just to get that one orientation bit. There's also a different way to load images using ImageIO that should allow you to read the metadata, but it too requires a fair bit of convoluted code.

Then there is this one odd post where someone says last month (without naming platform or JDK level or vendor) that ImageIO finally handles this correctly: https://stackoverflow.com/questions/47657302/does-imageio-read-take-exif-orientation-metadata-into-account

Since:
- this issue isn't hit often (first time in 5+ years as far as I know)
- there is an easy workaround
- fixing it will require lots of code that may introduce more bugs
- there's a chance that this is fixed in newer JDKs
then I am more inclined to either close this as 'no plan to fix' or wait and test some newer JDKs.

**Comment by niemela** [comment]
Maybe "fix" it by adding it to the documentation as an "known issue" with the workaround?

**Comment by johnbrvc** [comment]
IMHO, since no one asked, as someone who has only been in the graphics display business 25 or 30 years, ignoring a problem like this and hoping it will be fixed by a future library release is not the right approach.  Camera and display resolution is constantly getting better and better.  We're being asked to write display software for 8K displays now and doing image processing on various format images.  That being said, having extremely high resolution images (possibly taken in portrait mode) with whatever rotation the device capturing the image chooses to use must be supported by the software, in this case, the resolver (IMHO).

- Because this issue hasn't occurred in "5+ years" is poor reasoning.  Newer cameras/phones == higher resolution **by default**.  5+ years ago we probably wouldn't have _seen_ an image 4000x6000 as a team photo, although such capture devices did exist...
- The workaround is not "easy" - it is time consuming and can be sprung on the user as a surprise at the last minute (uh... ACPC 2022 for example).  Sitting on a dark porch 15 minutes before a presentation trying to determine which images come out rotated, then manually resizing them to "trick" the software into working is NOT a good workaround. (IMHO) 
- "Fixing it MAY [sic] require lots of code" - but software is supposed to make user's lives _easier_, not harder.  IMHO, this is a poor excuse for not fixing it.  Instead of all the elaborate library calls to re-read the image 2 or 3 times, how about just writing a routine that opens the image file and looks for the required rotation flags directly?  It may be a good interim solution and relatively small amount of code and processing power.  Of course, it won't support every image format known to man, but I think supporting the most common formats is a reasonable solution: jpg, png, bmp, tiff?  
- Waiting (hoping?) on possible library fixes - why is this even a consideration?  "Oh, by the way, to use the resolver, you'll need to install JDK 20" - we already need a special JDK installed to run the thing as it is, but now we'll need a "real new one"? No thanks.

I know all of our time is valuable and we don't have much extra of it to devote to other "projects".  But, IMHO, this issue is kinda important and a bug and _maybe_ should take precedence over bells and whistles - just saying (yeah, I know bells and whistles are a lot more fun to write code for...)

Did I use enough IMHO's through out this?  If not, feel free to add lots more. ;-)

**Comment by deboer-tim** [comment]
Hi John,

FWIW, I put a lot of weight in your comments that people tend to use the resolver under time pressure. OTOH, I feel like I need to say that at contests we too often fall into the trap of not testing things until right before we go onstage or 'we got those images way too late but let's try to make it work' without time to step back and see workarounds or other options. 

That said :), some further comments:
- This really is a JDK bug, and I've confirmed it has existed since at least 2004. Yes, other people may have hit this in the resolver before and never reported it, but that's a pretty low incidence rate.
- Re: "just writing a routine". The problem is that there is no rotation flag in the regular Java library. We'd need to include another library and code path that would read all images first, do the rotation if necessary, and save back to disk. That's not a very simple workaround, could cause regressions, and if a JDK ever fixes this we'd continue to pay the price of rotating the images unnecessarily. (FWIW, it looks like this is only for JPEGs - orientation is used so infrequently in other formats that even Chrome doesn't handle it)
- Yes, I try to work around JDK and even OS bugs when possible, because I agree that users don't care where the problem is and we should make things simpler for them. However, I was trying to be clear that in this case the risk + work vs reward is not looking like a great tradeoff to me.
- It's minor, but I'm going to disagree with a graphics expert on their image comment. :) The JPEG orientation flag was created for cameras that didn't have enough CPU/power to rotate on the device, so _every_ old vertical photo would have hit this problem. Newer/larger cameras can rotate on-device, use different formats, or it becomes more likely that someone will downscale or edit, all of which would reduce the odds of someone else hitting this.

In all likelihood, this is just one of the first times that someone took vertical photos of the teams _and_ nobody in the chain happened to do any post-processing of them.

**Comment by nickygerritsen** [comment]
Note that even recent iPhones still set this flag…
